tsl/src/hypercore/hypercore_handler.c

```diff
@@ -1787,6 +1789,87 @@ hypercore_tuple_delete(Relation relation, ItemPointer tid, CommandId cid, Snapsh
 	return result;
 }
 
+/*
+ * Decompress a segment that contains the row given by ctid.
+ *
+ * This function is typically called during an upsert (ON CONFLICT DO UPDATE),
+ * where the conflicting row points to a compressed segment that needs to be
+ * decompressed before the update can take place. This function is used to
+ * decompress that segment into a set of individual rows.
+ *
+ * Returns the number of rows in the segment that were decompressed, or 0 if
+ * the TID pointed to a regular (non-compressed) tuple. The TID of the
+ * de-compressed conflicting row is returned via "new_ctid".
+ */
+int
+hypercore_decompress_conflict_segment(Relation relation, const ItemPointer ctid,
+									  TupleTableSlot *slot, Snapshot snapshot, ItemPointer new_ctid)
+{
+	HypercoreInfo *hcinfo;
+	Relation crel;
+	TupleTableSlot *cslot;
+	ItemPointerData decoded_tid;
+	TM_Result result;
+	TM_FailureData tmfd;
+	int n_batch_rows = 0;
+	uint16 tuple_index;
+	bool should_free;
+
+	/* Nothing to do if this is not a compressed segment */
+	if (!is_compressed_tid(ctid))
+		return 0;
+
+	Assert(TTS_IS_ARROWTUPLE(slot));
+
+	hcinfo = RelationGetHypercoreInfo(relation);
+	crel = table_open(hcinfo->compressed_relid, RowExclusiveLock);
```

Comment:
You should be able to open this table once for the entire statement instead of opening it each time. There is no risk of deadlocks here since the lock on the "main" chunk tableoid prevents this, but it affects performance (not a lot though, this is not called very frequently and blocking is not expected).

Not sure if there is a convenient point where this table can be opened though, so feel free to ignore.